rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+its place. It is up to the master to reassign splits from the crashed worker to
+the new worker.
+
+To improve worker recovery time, workers will periodically write checkpoints of
+their iterators to directories named using their worker ids. When the restarted
+worker connects, the master will tell it which iterator checkpoints to recover
+from.
+
+We will read and write this state through a MasterState interface which can be
+implemented using various storage backends. For use cases that require fault
+tolerance, the user must configure a fault-tolerant MasterState, e.g. Spanner
+internally, Cloud Spanner in GCP, or etcd externally. If fault tolerance isn't
+required, the user could configure state to be held in memory only.
+
+#### Leadership Transfer
+
+The master writes state to journal files so that the state can be recovered on
+restart. It is possible that a new master could be brought up while the old
+master is still running. If we aren't careful, this could result in corruption
+of the journal as both masters try to write to it.
+
+Ideally we could rely on a distributed coordination service such as ZooKeeper.
+However, this would add a significant burden to users who don't have access to a
+ZooKeeper cluster, and it would also require adding a new dependency on a
+ZooKeeper client.
+
+What TensorFlow does have is a FileSystem API. We will leverage this API to
+perform leadership transfer as follows:
+
+1.  The first master will create a file named "master_seqno_0". If it
+    successfully creates the file, it will consider itself the leader.
+1.  The leader master will check every N milliseconds that the "master_seqno"
+    file it created still exists. If the file no longer exists, the master will
+    cease operation immediately.
+1.  When a master thinks it should be leader, it attempts to atomically rename
+    the master_seqno_n file to master_seqno_n+1. If this succeeds, the master
+    will wait (N + M) milliseconds, verify that its renamed file still exists,
+    and begin acting as leader. This gives the previous leader time to notice
+    the rename.
+
+The above scheme relies on rename being atomic so that two masters don't both
+succeed at renaming the same file. Users may opt to use a filesystem that
+doesn't support atomic rename, but they do so at the (unlikely) risk of two
+concurrently running masters thinking they are leader. Common filesystems such
+as Posix and HDFS support atomic rename.
+
+#### Caveats
+
+This section calls out caveats that users will need to be aware of when using
+the tf.data service.
+
+-   Due to the nature of dataset splitting, elements will not be processed in
+    the same order as they were in the pre-distributed dataset. If a dataset
+    relies on the order of the input files, the user's assumptions will be
+    violated when splitting causes each input worker to process only a subset of
+    the input files.
+-   If a dataset doesn't support splitting, it must be moved after the part of
```

Comment:
I assume "dataset doesn't support splitting"  means "if a particular dataset operation doesn't support splitting" here

Reply:
Correct. Updated for clarity, thanks.